### PR TITLE
doc: add timer HLD

### PR DESCRIPTION
**Pull request by dbkinder**
Review and add timer HLD document

Signed-off-by: David B. Kinder <david.b.kinder@intel.com>

### CODE REVIEW

File: doc/developer-guides/timer-hld.rst [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+How it works
+************
+
+When the system boots, we check that the hardware supports a LAPIC
+tsc-deadline timer by checking ``CPUID.01H:ECX.TSC_Deadline[bit 24]``.
+If support is missing, we output an error message and panic the
+hypervisor.  Otherwise, we register the timer interrupt callback that
+raises a timer software interrupt on each logical CPU and set the lapic
+timer mode to tsc-deadline timer mode by writing the locap APIC LVT
```
Comment (gvancuts) [inline doc/developer-guides/timer-hld.rst]: s/locap/local?

File: doc/developer-guides/timer-hld.rst [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+   };
+
+   struct hv_timer {
+      struct list_head node;      /* link-list for all timers */
+      int mode;                   /* timer mode: one-shot or periodic */
+      uint64_t fire_tsc;          /* tsc deadline to interrupt */
+      uint64_t period_in_cycle;   /* period of the periodic timer in units of TSC cycles */
+      timer_handle_t func;        /* callback when time reached */
+      void *priv_data;            /* callback function private data */
+   };
```
Comment (gvancuts) [inline doc/developer-guides/timer-hld.rst]: Some of the explanations above are a little different (more verbose) than what is in the header file. The ones here seem better, should we align the ones in the code to match the HLD?

File: doc/developer-guides/timer-hld.rst [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+There are three external timer APIs (also defined in ``timer.h``):
+
+.. code-block:: c
+
+   static inline void initialize_timer(
+      struct timer *timer,
```
Comment (gvancuts) [inline doc/developer-guides/timer-hld.rst]: s/timer/hv_timer ?

File: doc/developer-guides/timer-hld.rst [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+      timer_handle_t func,
+      void *priv_data,
+      uint64_t fire_tsc,
+      int mode,
+      uint64_t period_in_cycle);
+
+   int add_timer(struct timer *timer);
```
Comment (gvancuts) [inline doc/developer-guides/timer-hld.rst]: s/timer/hv_timer ?

File: doc/developer-guides/timer-hld.rst [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+   void del_timer(struct timer *timer);
```
Comment (gvancuts) [inline doc/developer-guides/timer-hld.rst]: s/timer/hv_timer ?

File: doc/developer-guides/timer-hld.rst [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+Before adding a timer, we must initialize the timer by calling
+``initialize_timer``. The processor generates a timer interrupt when the
+value of timer-stamp counter is greater than or equal to the
+``fire_tsc``.  If you want to add a periodic timer, you should also pass
+the period (in tsc cycles), otherwise, the period_in_cycle will be
```
Comment (gvancuts) [inline doc/developer-guides/timer-hld.rst]: Can you enclose period_in_cycle in double-backticks?

File: doc/developer-guides/timer-hld.rst [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+ignored. When the timer interrupt occurs, it will call the callback
+function func with parameter priv_data.
```
Comment (gvancuts) [inline doc/developer-guides/timer-hld.rst]: Can you enclose func in double-backticks?

File: doc/developer-guides/timer-hld.rst [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+The ``initialize_timer`` function only initializes the timer data
+structure.  It will not program the ``IA32_TSC_DEADLINE_MSR`` to
+generate the timer interrupt. If you want to generate a timer interrupt,
+you must call ``add_timer`` to add the timer to the per_cpu_timer
+timer_list.
+
+The hypervisor will chose the nearest expired timer on the timer_list
```
Comment (gvancuts) [inline doc/developer-guides/timer-hld.rst]: s/chose/choose

File: doc/developer-guides/timer-hld.rst [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+``fire_tsc``.  If you want to add a periodic timer, you should also pass
+the period (in tsc cycles), otherwise, the period_in_cycle will be
+ignored. When the timer interrupt occurs, it will call the callback
+function func with parameter priv_data.
+
+The ``initialize_timer`` function only initializes the timer data
+structure.  It will not program the ``IA32_TSC_DEADLINE_MSR`` to
+generate the timer interrupt. If you want to generate a timer interrupt,
+you must call ``add_timer`` to add the timer to the per_cpu_timer
+timer_list.
```
Comment (gvancuts) [inline doc/developer-guides/timer-hld.rst]: Can you enclose per_cpu_timer and timer_list in double-backticks?

File: doc/developer-guides/timer-hld.rst [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+The hypervisor will chose the nearest expired timer on the timer_list
+and program ``IA32_TSC_DEADLINE_MSR`` by writing its value to fire_ts.
+When the fire_tsc expires, it raises an interrupt whos callback raises a
```
Comment (gvancuts) [inline doc/developer-guides/timer-hld.rst]: s/whos/whose

File: doc/developer-guides/timer-hld.rst [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+software interrupt. The hypervisor handles the software interrupt before
+the VM reenters the guest. (Currently, the hypervisor doesn't use the
+timer except the console). The timer software interrupt handler will
```
Comment (gvancuts) [inline doc/developer-guides/timer-hld.rst]: What does that mean: "the hypervisor doesn't use the timer except the console"?

File: doc/developer-guides/timer-hld.rst [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+check each expired timer on its timer_list. Before calling the expired
+timer callback handler, it will remove the timer from its logical CPU
+timer_list.  After calling the timer callback handler, it will re-add
+the timer to the timer_list if its a periodic timer.
```
Comment (gvancuts) [inline doc/developer-guides/timer-hld.rst]: s/its/it's

**Comment by dbkinder** [comment]
@gvancuts This whole document should really just be doxygen comments in the source code and we use the API generating system we have in place.  @hwang37 has said to publish this now and they'll use doxygen later.  sigh.

File: doc/developer-guides/timer-hld.rst (hunk 3 above)
Comment (dbkinder) [reply doc/developer-guides/timer-hld.rst]: I copied this code right from the source tree, so the names should be correct.

Comment (gvancuts) [reply doc/developer-guides/timer-hld.rst]: That's odd, look at this line: https://github.com/projectacrn/acrn-hypervisor/blob/master/hypervisor/include/arch/x86/timer.h#L34

File: doc/developer-guides/timer-hld.rst (hunk 4 above)
Comment (gvancuts) [reply doc/developer-guides/timer-hld.rst]: https://github.com/projectacrn/acrn-hypervisor/blob/master/hypervisor/include/arch/x86/timer.h#L54

File: doc/developer-guides/timer-hld.rst (hunk 5 above)
Comment (gvancuts) [reply doc/developer-guides/timer-hld.rst]: https://github.com/projectacrn/acrn-hypervisor/blob/master/hypervisor/include/arch/x86/timer.h#L55

**Comment by gvancuts** [comment]
> @gvancuts This whole document should really just be doxygen comments in the source code and we use the API generating system we have in place. @hwang37 has said to publish this now and they'll use doxygen later. sigh.

I agree this should be self-contained as doxygen comments in the code.

File: doc/developer-guides/timer-hld.rst (hunk 5 above)
Comment (dbkinder) [reply doc/developer-guides/timer-hld.rst]: ah.  I see what happened.  For the functions, I didn't copy from the source code.  I just used what was in the doc.  That just goes to show you how the doc and code can get out of sync and why this really should be maintained as doxygen comments as soon as possible.

**Comment by gvancuts** [comment]
> ah. I see what happened. For the functions, I didn't copy from the source code. I just used what was in the doc. That just goes to show you how the doc and code can get out of sync and why this really should be maintained as doxygen comments as soon as possible.

Yes, this is bound to happen again at some point in time... moving this to Doxygen is the best option to make sure we don't have documents that become content-rotten :)

**Comment by dbkinder** [comment]
PM decision is to not publish this doc, and instead update the header files too add this API documentation (and publish in the API area).  @hwang37 Should we close this then, and add an issue to track?

**Comment by hwang37** [comment]
Yes, we can close this PR. I have a separate task to generate API 2.0 for all component, no need a new issue.

**Comment by dbkinder** [comment]
closed per PM decision

**Pull request closed without merging**